### PR TITLE
[dtensor][random] allow user to manual_seed different seed on device mesh; only sync RNG state in WORLD when manual_seed has not been called

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141532
* __->__ #141223
* #141220
* #141731

**Summary**
This PR proposes 4 changes to DTensor RNG management:
1. DTensor allows users to eagerly initialize the RNG tracker by calling `torch.distributed.tensor._random.manual_seed`.
2. DTensor `manual_seed` no longer checks the integrity of the `seed` argument. Users are responsible for setting the same seed on all ranks within an SPMD group, but if there are multiple separate SPMD groups (e.g. across pipeline stages), users should set a _different_ seed for each SPMD group. For cases like Pipeline Parallel, users can set different initial seed for pipelining stages by calling
```
world_mesh = init_device_mesh(
    device_type="cuda",
    mesh_shape=(2, 2, 2),
    mesh_dim_names=("pp", "dp", "tp"),
)
pp_mesh = world_mesh["pp"]
pp_rank = pp_mesh.get_local_rank()
spmd_mesh = world_mesh["dp", "tp"]._flatten("spmd")  # this flattening is only needed if you need to call collective over this mesh
torch.distributed.tensor._random.manual_seed(123+pp_rank, spmd_mesh)
```

In other word, if users want to call `torch.distributed.tensor._random.manual_seed`, they will be responsible for passing in the right value and DTensor won't perform any checks on it. If the current rank is not a part of the mesh, it will use the current device RNG state to initialize.

3. `OffsetBasedRNGTracker` still performs RNG state synchronization by broadcasting the RNG state on rank 0 to `WORLD`. However, calling `torch.distributed.tensor._random.manual_seed` is an exception. In this case, no broadcast will happen.

4. Enforce that the `manual_seed` call only accept "full mesh" i.e. the DTensor RNG state on every rank must be set through the call. This makes sure that no rank has its RNG state left uninitialized and the SPMD ranks have their RNG state synchronous.

**Motivation**
tl;dr

1. Lazily initializing DTensor RNG tracker causes hang in non-SPMD code such as Pipeline Parallel.
2. Users may want to set different seed on ranks in one device mesh.
3. We want to keep the old behavior if users prefer not curating the RNG state and want to have DTensor take care of it.

see detail in https://github.com/pytorch/pytorch/issues/140301

**Test**
`pytest test/distributed/_tensor/test_random_ops.py`
`pytest test/distributed/tensor/parallel/test_tp_random_state.py`

cc @wanchaol @tianyu-l @wz337 @d4l3k @H-Huang @awgu @kwen2501 @fegin @fduwjj @wconstab @c-p-i-occ @c-p-i-o

### CODE REVIEW

**Comment by wconstab** [comment]
looks pretty good to me.  I want to make a few changes to the PR desc for other users to understand better.

1. 
image
Update this code snippet to show how pp_device_mesh is defined.  (it should be a submesh of the world that includes all the dimensions _other than_ PP dim.  I would think the better variable name is `non_pp_mesh` or `spmd_mesh`

2. change this text 
image
Users are responsible for setting the same seed on all ranks within an SPMD group, but if there are multiple separate SPMD groups (e.g. across pipeline stages), users should set a _different_ seed for each SPMD group.

File: torch/distributed/tensor/_random.py [hunk 1]
```diff
@@ -71,20 +71,13 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
...
-            raise RuntimeError(
-                f"calling manual_seed function over {device_mesh} but received different seed values on ranks:",
-                f"seed on rank {dist.get_rank()} is {seed}, and seed on rank {rank} is {object}!",
-            )
     # instantiate a RNG tracker if haven't. By default DTensor uses an
     # OffsetBasedRNGTracker to perform random operators.
     global _rng_tracker
     if not _rng_tracker:
-        _rng_tracker = OffsetBasedRNGTracker(device_mesh.device_type)
+        _rng_tracker = OffsetBasedRNGTracker(
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:78]: what if a user calls manual seed a second time?  it looks like that would be a silent no-op, which is not expected.  should we remove the `if not _rng_tracker` check and just keep re-initializing new trackers each time manual_seed is called?  Or if there is some reason reinit is not supported, we should probably raise a RuntimeError here.

**Review by tianyu-l — COMMENTED** [review]
I agree with and like everything @wconstab said!

**Comment by wconstab** [comment]
@XilunWu FYI i got a PR on torchtitan side which uses this PR.  It crashes on something maybe not supported well inside OffsetBasedRNGTracker.  Maybe this is a regression after not using TensorParallelRNGTracker for TorchTitan?

Pytorch branch: this PR
Torchtitan branch: https://github.com/pytorch/torchtitan/pull/689 
Run command: `TORCH_LOGS=+pp CONFIG_FILE=./train_configs/debug_model.toml NGPU=8 LOG_RANK=0,1,2,3,4,5,6,7 ./run_llama_train.sh --job.dump_folder outputs/3d_compile --model.flavor debugmodel --experimental.pipeline_parallel_degree 2 --experimental.pipeline_parallel_split_points layers.4 --training.data_parallel_shard_degree 2 --training.tensor_parallel_degree 2`

Error:
<details>
```
[rank5]:[rank5]:     layer.init_weights()                                                                                                 16:46:46 [545/4895]
[rank5]:[rank5]:   File "/data/users/whc/torchtitan/torchtitan/models/llama/model.py", line 331, in init_weights                                             
[rank5]:[rank5]:     self.attention.init_weights(self.weight_init_std)                                                                                       
[rank5]:[rank5]:   File "/data/users/whc/torchtitan/torchtitan/models/llama/model.py", line 170, in init_weights                                             
[rank5]:[rank5]:     nn.init.trunc_normal_(linear.weight, mean=0.0, std=0.02)                                                                                
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/nn/init.py", line 224, in trunc_normal_                                                               
[rank5]:[rank5]:     return _no_grad_trunc_normal_(tensor, mean, std, a, b, generator=generator)                                                             
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/nn/init.py", line 47, in _no_grad_trunc_normal_                                                       
[rank5]:[rank5]:     tensor.uniform_(2 * l - 1, 2 * u - 1, generator=generator)                                                                              
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/_compile.py", line 32, in inner                                                                       
[rank5]:[rank5]:     return disable_fn(*args, **kwargs)                                                                                                      
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/_dynamo/eval_frame.py", line 721, in _fn                                                              
[rank5]:[rank5]:     return fn(*args, **kwargs)                                                                                                              
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/distributed/tensor/_api.py", line 343, in __torch_dispatch__                                          
[rank5]:[rank5]:     return DTensor._op_dispatcher.dispatch(                                                                                                 
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/distributed/tensor/_dispatch.py", line 211, in dispatch                                               
[rank5]:[rank5]:     with rng_context:                                                                                                                       
[rank5]:[rank5]:   File "/home/whc/.conda/envs/pytorch-3.10/lib/python3.10/contextlib.py", line 135, in __enter__                                            
[rank5]:[rank5]:     return next(self.gen)                                                                                                                   
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/distributed/tensor/_random.py", line 176, in _distribute_region                                       
[rank5]:[rank5]:     self._set_pre_op_offset(spec)                                                                                                           
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/distributed/tensor/_random.py", line 255, in _set_pre_op_offset                                       
[rank5]:[rank5]:     dim_map = spec.dim_map                                                                                                                  
[rank5]:[rank5]:   File "/data/users/whc/pytorch/torch/distributed/tensor/_dtensor_spec.py", line 164, in dim_map                                            
[rank5]:[rank5]:     raise ValueError(                                                                                                                       
[rank5]:[rank5]: ValueError: Tensor dim 0 is already sharded on mesh dim 0, DTensor operator implementation does not support things like hybrid sharding stra
tegies yet (i.e. [Shard(0), Shard(0)])  
```
</details>

File: torch/distributed/tensor/_random.py (hunk 1 above)
Comment (XilunWu) [reply torch/distributed/tensor/_random.py:78]: not really. This if-branch only skips the RNG tracker initialization part. The seed setting remains being executed.

**Comment by XilunWu** [comment]
successfully run test in https://github.com/pytorch/torchtitan/pull/689 after rebasing the PR on top.

**Comment by wconstab** [comment]
> spmd_mesh = world_mesh["dp", "tp"]._flatten("spmd")

Just curious: is this part necessary? or can we just pass `world_mesh["dp", "tp"]` into manual_seed?

> If the current rank is not a part of the mesh, it will use the current device RNG state to initialize.

This sounds wrong to me- if the user passes a mesh that is supposed to be the 'spmd mesh', and the current rank is not a part of that mesh, shouldn't we raise an error?

File: torch/distributed/tensor/_random.py [hunk 2]
```diff
@@ -71,20 +71,13 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
...
+            device_mesh.device_type, run_state_sync=False
+        )
 
     # the current rank is in mesh
     if device_mesh.get_coordinate() is not None:
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:83]: should we assert this instead of silent-no-op-ing it?

if we require users to pass in a device mesh that is defined as the 'spmd world', then it is wholly reasonable to make it error out if the mesh does not contain our rank.

**Review by wconstab — APPROVED** [review]
approving as long as you add the runtime error in manual_seed for missing ranks.

File: torch/distributed/tensor/_random.py [hunk 3]
```diff
@@ -52,7 +52,10 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
 
     Args:
         seed (int): The desired seed.
-        device_mesh (:class:`DeviceMesh`): The device mesh to set the seed.
+        device_mesh (:class:`DeviceMesh`): The device mesh to set the seed. It is
+            required that the ``device_mesh`` include the calling rank. This is
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:56]: nit: "include the calling rank" is too weak a statement.  

It is required that 'device_mesh' contains the superset of all meshes that will be used with this DTensor.  It would be good to define 'spmd world' and then say that this mesh must be the spmd_world.
